Common BDN code refactor

adamsitnik opened the pull request:
I was not sure how to name the project, `BechmarkDotNet.Extensions` was my best idea.

jorive commented on src/benchmarks/micro/MicroBenchmarks.csproj:
```diff
@@ -32,6 +30,10 @@
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\harness\BenchmarkDotNet.Extensions\BenchmarkDotNet.Extensions.csproj" />
```
Maybe the PATH could be something like `..\..\BenchmarkDotNet\Extensions\Extensions.csproj`?

adamsitnik replied:
I am not sure if this is a better idea. I don't think that we are ever gonna need more `BenchmarkDotNet` projects but I think that we might add some other harnesses in the future.

jorive commented on src/benchmarks/micro/Program.cs:
```diff
@@ -18,24 +14,9 @@ class Program
                 .FromAssembly(typeof(Program).Assembly)
-                .Run(args, GetConfig())
+                .Run(args, RecommendedConfig.Create(
+                    artifactsPath: new DirectoryInfo(Path.Combine(Path.GetDirectoryName(typeof(Program).Assembly.Location), "BenchmarkDotNet.Artifacts")), 
```
This should probably be moved to the `RecommendedConfig` instead.

adamsitnik replied:
the problem is that we want `typeof(Program).Assembly.Location` which is different than `BenchmarkDotNet.Extensions.dll` location

We could use `Assembly.GetEntryAssembly` or ``Assembly.GetExecutingAssembly` but I have bad experience using those methods (they have platform specific behaviour depending on what has actually run the benchmarks) and AFAIR they are not implemented on CoreRT (which I hope to add sooon)

jorive commented on src/benchmarks/real-world/Microsoft.ML.Benchmarks/Microsoft.ML.Benchmarks.csproj:
```diff
@@ -1,15 +1,18 @@
   </ItemGroup>
 
+  <ItemGroup>
+      <ProjectReference Include="..\..\..\harness\BenchmarkDotNet.Extensions\BenchmarkDotNet.Extensions.csproj" />
```
w.r.t. The referenced project PATH please see above.

jorive commented on src/benchmarks/real-world/Microsoft.ML.Benchmarks/Program.cs:
```diff
@@ -25,26 +20,11 @@ class Program
                 .FromAssembly(typeof(Program).Assembly)
-                .Run(args, GetConfig())
+                .Run(args, RecommendedConfig.Create(
+                    artifactsPath: new DirectoryInfo(Path.Combine(Path.GetDirectoryName(typeof(Program).Assembly.Location), "BenchmarkDotNet.Artifacts")), 
```
Same as above.

adamsitnik replied:
same as above (`typeof(Program).Assembly` != `BenchmarkDotNet.Extensions.dll`)

jorive commented on src/harness/BenchmarkDotNet.Extensions/BenchmarkDotNet.Extensions.csproj:
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Should we add:
```xml
<UseSharedCompilation>false</UseSharedCompilation>
<GenerateDocumentationFile>False</GenerateDocumentationFile>
<TreatWarningsAsErrors>True</TreatWarningsAsErrors>
<WarningLevel>4</WarningLevel>
<SuppressNETCoreSdkPreviewMessage>True</SuppressNETCoreSdkPreviewMessage>
<SignAssembly>false</SignAssembly>
```

adamsitnik replied:
I added the reference to `common.props` file

jorive commented:
These package versions could be updated:
```diff
-    <PackageReference Include="Jil" Version="2.15.4" />
+    <PackageReference Include="Jil" Version="2.16.0" />
-    <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
-    <PackageReference Include="protobuf-net" Version="2.3.7" />
-    <PackageReference Include="System.Memory" Version="4.5.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.1" />
+    <PackageReference Include="protobuf-net" Version="2.4.0" />
+    <PackageReference Include="System.Memory" Version="4.5.2" />
```

adamsitnik commented:
> These package versions could be updated:

Yes, but it could affect the reported time. For example update of Newtonsoft.Json from 11 to 12 can have some perf impact.  We are now testing the stability of the reported results and I would prefer to not update them now.

adamsitnik commented:
@jorive PTAL at the changes that I have introduced in https://github.com/dotnet/performance/pull/341/commits/ee7d5c7eb204d6571defbb058cee696cecf9fcf5

the CI has failed with some strange error after I have introduced a new project:

```log
$ pushd "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\artifacts\bin\BenchmarkDotNet.Extensions\Release\netcoreapp3.0"
Searching BenchmarkDotNet output files with: BenchmarkDotNet.Artifacts/**/*-full.json
$ popd
$ pushd "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\artifacts\bin"
$ popd
Traceback (most recent call last):
  File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchmarks_ci.py", line 250, in <module>
    __main(sys.argv[1:])
  File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchmarks_ci.py", line 245, in __main
    benchview.run_scripts(args, verbose, BENCHMARKS_CSPROJ)
  File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchview.py", line 454, in run_scripts
    jobtype=args.benchview_run_type
  File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchview.py", line 167, in submission
    raise ValueError("No `measurement.json` were specified.")
ValueError: No `measurement.json` were specified.
```

we tried to search for the results in `bin\BenchmarkDotNet.Extensions\Release\netcoreapp3.0` instead `
bin\MicroBenchmarks\Release\netcoreapp3.0`

I debugged it and it turned out that we were searching for the results in the **first** folder that matched `'**/{Configuration}/{TargetFramework}` 

I have changed it to `**/{ProjectName}/{Configuration}/{TargetFramework}`.

jorive commented:
> @jorive PTAL at the changes that I have introduced in ee7d5c7
> the CI has failed with some strange error after I have introduced a new project:
> $ pushd "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\artifacts\bin\BenchmarkDotNet.Extensions\Release\netcoreapp3.0"
> Searching BenchmarkDotNet output files with: BenchmarkDotNet.Artifacts/**/*-full.json
> $ popd
> $ pushd "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\artifacts\bin"
> $ popd
> Traceback (most recent call last):
>   File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchmarks_ci.py", line 250, in <module>
>     __main(sys.argv[1:])
>   File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchmarks_ci.py", line 245, in __main
>     benchview.run_scripts(args, verbose, BENCHMARKS_CSPROJ)
>   File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchview.py", line 454, in run_scripts
>     jobtype=args.benchview_run_type
>   File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchview.py", line 167, in submission
>     raise ValueError("No `measurement.json` were specified.")
> ValueError: No `measurement.json` were specified.
> 
> we tried to search for the results in bin\BenchmarkDotNet.Extensions\Release\netcoreapp3.0 instead  bin\MicroBenchmarks\Release\netcoreapp3.0
> I debugged it and it turned out that we were searching for the results in the first folder that matched '**/{Configuration}/{TargetFramework}
> I have changed it to **/{ProjectName}/{Configuration}/{TargetFramework}.

Sure, I'll take a look.

adamsitnik commented:
This has blocked me in #352. I am going to merge this now to unblock #352 and allow for ML.NET per runs.

jorive commented:
@adamsitnik Did you solve/fix the exception you were getting?

adamsitnik commented:
@jorive yes, I was able to repro it locally, fix and test
